design/drive-status.md

```diff
@@ -0,0 +1,129 @@
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# drive-status
+
+## Status
+
+provisional
+
+## Table of Contents
+
+   * [Title](#title)
+      * [Status](#status)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints-optional)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Design Details](#design-details)
+         * [Work Items](#work-items)
+         * [Dependencies](#dependencies)
+         * [Test Plan](#test-plan)
+         * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+         * [Version Skew Strategy](#version-skew-strategy)
+      * [References](#references)
+
+## Summary
+
+This document explains how drives information, including health data, will be
+retrieved, collected and monitored to provide constant up-to-date status on
+drives present in a server.
+
+## Motivation
+
+We need to be able to collect and update enough information about drives in a
+server to monitor their status and identify them for maintenance and
+replacement.
+Valuable information are for example the current status, type, serial number,
+bay location, part number.
+We need to be able to refresh the information about the status of the drives
+and, if a drive has been replaced, we need to be able to retrieve the
+new information and update the hardware inventory to provide a constant
+up-to-date vision on the status of the drives.
+
+### Goals
+
+1. Decide which information on the drives to retrieve
+2. Agree on how to retrieve such information
+3. Provide a way to manually/automatically update the information
+
+### Non-Goals
+
+N/A
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+- For each drive we need to collect at least the following information:
+    * Capacity
+    * Location
+    * Manufacturer
+    * Model
+    * Part Number
+    * Serial Number
+    * Status
+    * SMART details
+    
+- NVMe drives require special tools to get SMART data, although
+[smartmontools](https://www.smartmontools.org/wiki/NVMe_Support) supports NVMe
+devices since version 6.5
+ 
+- [Hdparm](https://en.wikipedia.org/wiki/Hdparm) can provide SMART data for
+most of drives
+
+- [python-hardware](https://github.com/redhat-cip/hardware) can collect SMART
+data
+
+[Ironic Python Agent (IPA)](https://github.com/openstack/ironic-python-agent)
+is already able to collect most of this information.
+Using a subset of the IPA code running inside a container in the hosts where we
+want to collect the drives data from can also allow to provide periodic
+updates on all the required info.
+This new service will talk to inspector that needs to be able to publish
+up-to-date data at any time.
+
+### Risks and Mitigations
+
+N/A
```

Comment:
There is some risk in having this approach rejected by the upstream Ironic team, isn't there?